// rxjs-spy/plugin/log-plugin.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var logger_1 = require("../logger");
var match_1 = require("../match");
var plugin_1 = require("./plugin");
var util_1 = require("../util");
var LogPlugin = (function (_super) {
    __extends(LogPlugin, _super);
    function LogPlugin(spy, match, partialLogger) {
        if (partialLogger === void 0) { partialLogger = logger_1.defaultLogger; }
        var _this = _super.call(this, "log(" + match_1.toString(match) + ")") || this;
        _this.auditor_ = spy.auditor;
        _this.logger_ = logger_1.toLogger(partialLogger);
        _this.match_ = match;
        return _this;
    }
    LogPlugin.prototype.beforeComplete = function (ref) {
        this.log_(ref, "complete");
    };
    LogPlugin.prototype.beforeError = function (ref, error) {
        this.log_(ref, "error", error);
    };
    LogPlugin.prototype.beforeNext = function (ref, value) {
        this.log_(ref, "next", value);
    };
    LogPlugin.prototype.beforeSubscribe = function (ref) {
        this.log_(ref, "subscribe");
    };
    LogPlugin.prototype.beforeUnsubscribe = function (ref) {
        this.log_(ref, "unsubscribe");
    };
    LogPlugin.prototype.log_ = function (ref, notification, param) {
        var _this = this;
        var _a = this, auditor_ = _a.auditor_, match_ = _a.match_;
        if (match_1.matches(ref, match_)) {
            auditor_.audit(this, function (ignored) {
                var logger_ = _this.logger_;
                var observable = ref.observable, subscriber = ref.subscriber;
                var tag = match_1.read(observable);
                var type = util_1.inferType(observable);
                var matching = (typeof match_ === "string") ? "" : "; matching " + match_1.toString(match_);
                var audit = (ignored === 0) ? "" : "; ignored " + ignored;
                var description = tag ?
                    "Tag = " + tag + "; notification = " + notification + matching + audit :
                    "Type = " + type + "; notification = " + notification + matching + audit;
                switch (notification) {
                    case "error":
                        logger_.error(description + "; error =", param);
                        break;
                    case "next":
                        logger_.log(description + "; value =", param);
                        break;
                    default:
                        logger_.log(description);
                        break;
                }
            });
        }
    };
    return LogPlugin;
}(plugin_1.BasePlugin));
exports.LogPlugin = LogPlugin;
